### DISCUSSION TITLE
Build CUDA aware MOOSE - Issues to buils PETSc - SUPERLU_DIST

### DISCUSSION THREAD

**Discussion by DanoLV** [idaholab/moose#27694] [Q&A Getting Started]
### Check these boxes if you have followed the posting rules.

- [X] Q&A Getting Started is the most appropriate category for my question (trouble installing, beginner user, ...)
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (avoid screenshots if possible, triple back quotes before/after pasted text, etc)

### Issue or question about MOOSE

Hello. 
I have been trying to build MOOSE with CUDA aware OpenMPI an PETSc, but my efforts have been unsuccessful until now.
OpenMPI is CUDa aware in my build, but when I try to build PETSc it fails to make SUPERLU_DIST.
![image](https://github.com/idaholab/moose/assets/59875578/3043345d-c893-4958-8d33-79b23a03f37c)

Configure_petsc.sh is been called wth the options:

python3 ./configure --with-64-bit-indices \
      --with-cxx-dialect=C++17 \
      --with-debugging=no \
      --with-fortran-bindings=0 \
      --with-mpi=1 \
      --with-openmp=1 \
      --with-cuda=1 \
      --with-shared-libraries=1 \
      --with-sowing=0 \
      --download-fblaslapack=1 \
      --download-hypre=1 \
      --download-metis=1 \
      --download-mumps=1 \
      --download-ptscotch=1 \
      --download-parmetis=1 \
      --download-scalapack=1 \
      --download-slepc=1 \
      --download-strumpack=1 \
      --download-superlu_dist=1 \
      $HDF5_STR \
      $MUMPS_ARM_STR \
      $MAKE_NP_STR \
      "$@"

I atach the PETSc config.log.
[configure.log](https://github.com/idaholab/moose/files/15408876/configure.log)

Thank in advance.

### Diagnostics for MOOSE installation issues

[Optional] Output of the diagnostics scripts. To run the script: cd ~/projects/moose/scripts; ./diagnostics.sh

**Comment by GiudGiud** [comment]
Hello

You could remove superlu-dist by manually editing the update and rebuild petsc script

Guillaume

**Comment by lindsayad** [comment]
@balay I don't know too much about this, but looking in the `configure.log`, I see
```
/home/dano/discos/datos/proyects/moose/petsc/arch-moose/externalpackages/git.superlu_dist/SRC/pdgstrs_lsum_cuda.cu(1746): error: no instance of overloaded function "atomicAdd" matches the argument list
```
My top google hit for this was [this stack overflow post](https://stackoverflow.com/questions/75565347/atomicadd-showing-error-no-instance-of-overloaded-function-atomicadd-mathes) which suggests this error might be fixable with flags. Have you run into this before?

**Reply by jczhang07** [reply]
From configure.log

```
/home/dano/discos/datos/proyects/moose/petsc/arch-moose/externalpackages/git.superlu_dist/SRC/pdgstrs_lsum_cuda.cu(1281): error: no instance of overloaded function "atomicAdd" matches the argument list
            argument types are: (double *, double)
         temp=atomicAdd(&lsum[il+irow + j*iknsupc],-temp1);
```

From a code comment in PETSc
`  The 64-bit floating-point version of atomicAdd() is only supported by devices of compute capability 6.x and higher.`

From configure.log, I found petsc failed to detect the cuda arch of the system

```
Checking for program /usr/local/cuda/extras/demo_suite/deviceQuery...not found
    Checking for program /home/dano/discos/datos/proyects/moose/petsc/lib/petsc/bin/win32fe/deviceQuery...not found
  Unable to find programs: ['deviceQuery'] in listing of the specific search path: ['/usr/local/cuda/extras/demo_suite']
      Warning accessing /usr/local/cuda/extras/demo_suite gives errors: [Errno 2] No such file or directory: '/usr/local/cuda/extras/demo_suite'
Executing: nvcc -c -o /tmp/petsc-n5_n6b3o/config.packages.cuda/conftest.o -I/tmp/petsc-n5_n6b3o/config.setCompilers -I/tmp/petsc-n5_n6b3o/config.types -I/tmp/petsc-n5_n6b3o/config.packages.cuda  -ccbin mpicxx -std=c++20 -Xcompiler -fPIC -Xcompiler -fvisibility=hidden -O3  /tmp/petsc-n5_n6b3o/config.packages.cuda/conftest.cu 
Successful compile:
Source:
#include "confdefs.h"
#include "conffix.h"
#include <stdio.h>
                    #include <cuda_runtime.h>
                    #include <cuda_runtime_api.h>
                    #include <cuda_device_runtime_api.h>

int main(void) {
int cerr;
                cudaDeviceProp dp;
                cerr = cudaGetDeviceProperties(&dp, 0);
                if (cerr) printf("Error calling cudaGetDeviceProperties\n");
                else printf("%d\n",10*dp.major+dp.minor);
                return(cerr);
  return 0;
}

Executing: nvcc  -o /tmp/petsc-n5_n6b3o/config.packages.cuda/conftest   -ccbin mpicxx -std=c++20 -Xcompiler -fPIC -Xcompiler -fvisibility=hidden -O3 /tmp/petsc-n5_n6b3o/config.packages.cuda/conftest.o 
Testing executable /tmp/petsc-n5_n6b3o/config.packages.cuda/conftest to see if it can be run
Executing: /tmp/petsc-n5_n6b3o/config.packages.cuda/conftest
Running Executable with threads to time it out at 60
Executing: /tmp/petsc-n5_n6b3o/config.packages.cuda/conftest
stdout: Error calling cudaGetDeviceProperties
ERROR while running executable: Could not execute "['/tmp/petsc-n5_n6b3o/config.packages.cuda/conftest']":
Error calling cudaGetDeviceProperties
```

I don't know the exact reason why the checking failed.  Perhaps @DanoLV should double check if the GPU is new enough to support cuda arch 60,  or explicitly set that in petsc configure options,  e.g., `--with-cuda-arch=70`?

**Reply by jczhang07** [reply]
@DanoLV I added https://gitlab.com/petsc/petsc/-/merge_requests/7572 off `petsc/release` to get more info on the failure of detecting cuda arch.  You might want to have a try.

**Reply by DanoLV** [reply]
Thaks you all for the comments. 
All this stuff is pretty new for my, so I do not really get all the details. 

What I can tell is that I'm using a GeForce GTX 1080, which has compute capability 6.1. So it meets the requirements.

Finally, I did manage to get a complete build of MOOSE that works!
To make it work, I installed CUDA 12.5, HDF5 1.14.4, UCX 1.16.0,  and  OpenMPI 5.0.3 first. Then, downloaded MOOSE and changed the PETSc vesion. When the script update_and_rebuild_petsc.sh gets PETSc from github, it downloads version 3.20. I replaced it with version 3.21.1 and commented the lines from the script to avoid it beening replaced again. Afted that I did not get another errors in the build proccess.

I set the scrip configure_petsc.sh to run with these options:
  python3 ./configure --with-64-bit-indices \
      --with-cxx-dialect=C++17 \
      --with-debugging=no \
      --with-fortran-bindings=0 \
      --with-mpi=1 \
      --with-openmp=1 \
      --with-cuda=1 --with-cuda-dialect=cxx17 -CUDAOPTFLAGS="" -Wno-deprecated-gpu-targets --download-kokkos --download-kokkos-kernels\
      --with-shared-libraries=1 \
      --with-sowing=0 \
      --download-fblaslapack=1 \
      --download-hypre=1 \
      --download-metis=1 \
      --download-mumps=1 \
      --download-ptscotch=1 \
      --download-parmetis=1 \
      --download-scalapack=1 \
      --download-strumpack=1 --download-slate\
      --download-slepc=1 \
      --download-superlu_dist=1 \
      $HDF5_STR \
      $MUMPS_ARM_STR \
      $MAKE_NP_STR \
      "$@"

After that, I continued with the steps explained in https://mooseframework.inl.gov/getting_started/installation/hpc_install_moose.html#!

The run_tests execution gives several error, but it seems to be due to the lack of definition of -vec_type -mat_type in the tests definitions. I have to do more checks on this matter.

**Reply by milljm** [reply]
Thank you for posting. I've been working on getting a Cuda variant Conda package set made available to the masses, and your input helps.

With older PETSc versions I was being stopped by several PETSc contribs. But your configure line indicates these contribs are now working properly.

**Reply by lindsayad** [reply]
I think that any tests using boomeramg will likely fail since boomeramg always assumes the vec and mat will be on the device but by default we just create those on the host

**Reply by Traiwit** [reply]
Hi @DanoLV

could you please share your `update_and_rebuild_petsc.sh` and `configure_petsc.sh`? i'm having hard time running update_and_rebuild_petsc.sh as well

also did you install CUDA 12.5, HDF5 1.14.4, UCX 1.16.0, and OpenMPI 5.0.3 to your linux or conda environment? 

I'm super new to this compile PETSC and GPU stuff, I kept running into error after error, could you please guide me?

Thank you for you help!

Kind regards,
Traiwit

**Reply by lindsayad** [reply]
They gave a pretty detailed configure line. Browsing it, I believe the additional arguments you would pass to `scripts/update_and_rebuild_petsc.sh` if you want to mimic their configuration are:
```
scripts/update_and_rebuild_petsc.sh --with-cxx-dialect=C++17 --with-cuda=1 --with-cuda-dialect=cxx17 -CUDAOPTFLAGS="" -Wno-deprecated-gpu-targets --download-kokkos --download-kokkos-kernels
```

**Reply by lindsayad** [reply]
make sure that you've configured openmpi with `--with-cuda`

**Reply by balay** [reply]
> /home/dano/discos/datos/proyects/moose/petsc/arch-moose/externalpackages/git.superlu_dist/SRC/pdgstrs_lsum_cuda.cu(1746): error: no instance of overloaded function "atomicAdd" matches the argument list

This issue was fixed by https://gitlab.com/petsc/petsc/-/merge_requests/7288/diffs?commit_id=cdfa477dea3fec089d157123d8a6f983b5ff4346

but yeah - generall - best to use latest version for GPU usage
